dune describe-project

emillon opened the pull request:
This new command describes what is available in a project.

There's not much going on here - that's to get the discussion going. We can start with just the public elements, and then extend to the private ones. A similar command will be useful to describe workspaces.

avsm commented:
This is extremely cool for big duniverses!  One early question: it is always awkward parsing this sort of CLI output (wrapping lines, escaping, etc) from scripts.  It might be nice to have a JSON alternative for output so that we can use jq to easily extract particular fields for scripting purposes.

rizo commented:
In addition this would potentially help with odoc integration. Both for simple dune projects and for duniverses. If the output for `dune describe` contained all the packages and their metadata, we could move the HTML generation logic to odoc.

emillon commented:
Hi! This is still alive :)

I just rebased and added JSON output. My box-fu is pretty weak, so the formatting code is not ideal, but I think this demonstrates the concept.

diml commented:
I can help with the formatting stuff, I'm (almost) comfortable with it :)

However, moving forward could you try to use the `Pp` API? We'll eventually get rid of `Format` and `Fmt` in Dune. The `Pp` API is safer and more functional.

emillon commented:
I started porting JSON formatting to `Pp` output. It's not yet what I'm looking for (I'm aiming for the `jq` output`) but I'm making progress. What's the idiomatic way to print in color? `User_message` directly?

diml commented:
The idea behind user message styles is to keep the styles abstract up to the point where they are actually rendered. In the future, we could imagine allowing the user the configure the concrete ANSI styles associated to these abstract styles via the user configuration file (`~/.config/dune/config`). If we want the user to be able to control the styles in the json output as well, then I guess it makes sense to use these styles directly.

BTW, one adds styles to a `Pp.t` via `Pp.tag`.  Tags are typed and it is possible to map the tags via `Pp.map_tags`, so you could even use some json specific tags while you do the json rendering.

emillon commented:
Now with colors and sexp output!

diml commented:
\o/

diml commented:
@emillon what's left to do on this PR?

One suggestion: what do you think about using just `dune describe` rather than `dune describe-project`?

emillon commented:
Describing a project is a bit different from describing a workspace, since in a workspace we probably only need to care about public things, so I was thinking to have a separate `describe-workspace` command as well. That could be solved by flags as well, but that might be a bit clunky.

Otherwise, left to do:
- [ ] display tests
- [ ] check if some other stanzas need printing
- [ ] (how) should we version the machine readable output (`describe --json` is bound to be used by external tools)
- [ ] Move JSON to its module
- [ ] Test & complete dyn/json conversion
- [ ] Test & complete dyn/dunelang conversion
- [ ] Test JSON coloring

diml commented:
I see the idea. Although having two different commands to ask Dune to describe things feels a bit much to me. It seems better to me to have a single command that does something sensible for humans by default, and command line options to get the fully defined and versioned behaviour for tools.

How about describing the project where Dune is run from in more details, and displaying only the public elements of other ones? Another idea would be to distinguish on vendored/non-vendored.

emillon commented:
Some ideas from a previous discussion:
- in order to wait a bit until the output is versioned, we could emit a message when the output is being piped (that it what `apt` does). Once/if we have settled on a format, it can be passed as a command-line argument
- `esy` has a similar command that describes a project - we should see what it's doing.

emillon commented:
This feature now exists.

emillon commented:
This feature now exists.
